TaskWithClass/MyString.cpp

```diff
@@ -0,0 +1,254 @@
+#include "pch.h"
+#include "MyString.h"
+
+const char* String::c_str() const noexcept
+{
+	return m_str;
+}
+
+size_t String::length() const
+{
+	return m_len;
+}
+
+size_t String::size() const
+{
+	return m_len;
+}
+
+size_t String::capacity() const noexcept
+{
+	return m_capacity;
+}
+
+void String::resize(size_t n, char c)
+{
+	if (m_len > n)
+	{
+		m_str[n] = '\0';
+		this->m_len = n;
+	}
+	else if (m_len < n)
+	{
+		char* mem = new char[n + 1];
+		memcpy(mem, m_str, m_len);
+		for (size_t i = m_len; i <= n - 1; i++)
+		{
+			mem[i] = c;
+		}
+		mem[n] = '\0';
+		delete[] m_str;
+		this->m_str = mem;
+		if (c != '\0')
+		{
+			this->m_len = n;
+		}
+		this->m_capacity = n + 1;
+	}
+}
+
+String& String::append(const String& str)
+{
+	if (str.length() == 0)
+	{
+		return *this;
+	}
+	size_t len = this->length() + str.length();
+	char* mem = new char[len + 1];
+	this->copy(mem, this->length());
+	char* start = &mem[this->length()];
+	str.copy(start, str.length());
+	mem[len] = '\0';
+	delete[] m_str;
+	this->m_str = mem;
+	this->m_len = len;
+	this->m_capacity = len + 1;
+	return *this;
+}
+
+String& String::operator=(const String& str)
+{
+	if (str.length() == 0)
+	{
+		return *this;
+	}
+	size_t len = this->length() + str.length();
+	char* mem = new char[len + 1];
+	this->copy(mem, this->length());
+	char* start = &mem[this->length()];
+	str.copy(start, str.length());
+	mem[len] = '\0';
+	delete[] m_str;
+	this->m_str = mem;
+	this->m_len = len;
+	this->m_capacity = len + 1;
+	return *this;
+}
+
+size_t String::copy(char* s, size_t len, size_t pos) const
+{
+	if (pos >= this->m_len)
+	{
+		throw std::out_of_range("invalid string position");
+	}
+	char* start = &this->m_str[pos];
+	if (m_len < len + pos)
+	{
+		len = m_len;
+	}
+	std::memcpy(s, start, len - pos);
```

Comment:
Add this test 

	String s = "abcd";
	std::string what;

	char a[256] = { 0, };
	s.copy(a, 0, 1);

This causes len - pos = 0 - 1 = -1 in this line. Since there's size_t in memcpy it will cause an epic crash.